language-support/hs/bindings/src/DA/Ledger.hs

```diff
@@ -59,5 +54,67 @@ getTransactionsPF lid party = do
     let req2 = GetTransactionsRequest lid now         Nothing    filter verbose
     stream <- getTransactions req1
     future <- getTransactions req2
-    past <- liftIO $ streamToList stream
+    past <- withTimeout $ liftIO $ streamToList stream
     return PastAndFuture { past, future }
+
+
+withGetTransactions
+    :: GetTransactionsRequest
+    -> (Stream [Transaction] -> LedgerService a)
+    -> LedgerService a
+withGetTransactions req act = do
+    stream <- getTransactions req
+    res <- mapIOLedgerService try (act stream)
+    liftIO $ closeStream stream EOS
+    case res of
+        Left e -> liftIO $ throwIO (e::SomeException)
+        Right x -> return x
+
+
+withGetTransactionTrees
+    :: GetTransactionsRequest
+    -> (Stream [TransactionTree] -> LedgerService a)
+    -> LedgerService a
+withGetTransactionTrees req act = do
+    stream <- getTransactionTrees req
```

Comment:
Once you have the `MonadUnliftIO` instance, you should be able to simplify this to
`bracket (getTransactionTrees req) (\s -> closeStream s EOS) (act stream)` which makes sure that the stream is closed even in the presence of asynchronous exceptions.

As it is this code will not close the stream if you get an exception after acquiring it but before you get to the `try` or after the `try`.

Reply:
yup

Reply:
Done. Using the `bracket` from `MonadLiftIO`.